### ISSUE TITLE
The "catch" block is not catching the error with one transaction using chain promises

### ISSUE THREAD

**Issue by jdnichollsc** [sequelize/sequelize#6409]
## What you are doing?

[issue-6409.zip](https://github.com/sequelize/sequelize/files/561895/issue-6409.zip)

I'm using Promises in one transaction.

With my following code, the error is **silent** and the **"catch"** is not executed.

``` javascript
var Q = require('q');
var models = require("../models");

var forms = require('./data/forms.json').forms || [];

var createField = function(field, formDB){

    return models.field.findOrCreate({
        where: {
            name: field.apiName,
        },
        defaults: {
            type: field.typeField,
            keyItem: field.keyItem,
            parentId: field.parentId
        }
    }).then(function(fieldDB){
        fieldDB = fieldDB[0] || fieldDB;
        return models.formField.findOrCreate({
            where: {
                formId: formDB.formId,
                fieldId: fieldDB.fieldId,
            },
            defaults: {
                customName: field.customName,
                isVisible: field.isVisible,
                isRequired: field.isRequired,
                position: field.position
            }
        });
    });
};
var createForm = function(form){
    return models.form.findOrCreate({
        where: {
            name: form.name
        }
    }).then(function(formDB){
        formDB = formDB[0] || formDB;
        return Q.all(form.fields.map(function(field){
            return createField(field, formDB);
        }));
    });
};

models.sequelize.sync().then(function () {

    var result = Q();
    forms.forEach(function (form) {
        result = result.then(function(){
            return createForm(form);
        });
    });

    result.then(function(){
        console.log("Success!");
    }).catch(function(err){
        console.log("Error!"); //********************** Doesn't work ***************************
        var msg = err;
    });
});
```

But using **transactions**, the **"catch"** is executed **(Why? I don't have any idea about that)**:

``` javascript
var Q = require('q');
var models = require("../models");

var forms = require('./data/forms.json').forms || [];

var createField = function(field, formDB, t){

    return models.field.findOrCreate({
        where: {
            name: field.apiName,
        },
        defaults: {
            type: field.typeField,
            keyItem: field.keyItem,
            parentId: field.parentId,
        },
        transaction: t
    }).then(function(fieldDB){
        fieldDB = fieldDB[0] || fieldDB;
        return models.formField.findOrCreate({
            where: {
                formId: formDB.formId,
                fieldId: fieldDB.fieldId,
            },
            defaults: {
                customName: field.customName,
                isVisible: field.isVisible,
                isRequired: field.isRequired,
                position: field.position
            },
            transaction: t
        });
    });
};

var createForm = function(form){
    return models.sequelize.transaction(function (t1){ 
        return models.form.findOrCreate({
            where: {
                name: form.name
            },
            transaction: t1
        }).then(function(formDB){
            formDB = formDB[0] || formDB;
            return models.sequelize.transaction(function (t2){
                return Q.all(form.fields.map(function(field){
                    return createField(field, formDB, t2);
                }));
            });
        });
    });
};

models.sequelize.sync().then(function () {

    var result = Q();
    forms.forEach(function (form) {
        result = result.then(function(){
            return createForm(form);
        });
    });

    result.then(function(){
        console.log("Success!");
    }).catch(function(err){
        console.log("Error!"); //******************** Works perfect! ***************************
        var msg = err;
    });
});
```
## What do you expect to happen?

The **catch** should be executed with one transaction.
## What is actually happening?

The **"catch"** is not executed and therefore the error is silent.

**Output (The error when the **catch** is executed)**

```
{
    "name": "SequelizeUniqueConstraintError",
    "message": "Validation error",
    "errors": [{
        "message": "custom_name must be unique",
        "type": "unique violation",
        "path": "custom_name",
        "value": "Site"
    }],
    "fields": {
        "custom_name": "Site"
    }
}
```

**Dialect:** postgres
**Database version:** 9.5.3
**Sequelize version:** 3.23.6

Thanks in advance, Nicholls

**Comment by felixfbecker** [comment]
Does the same happen when you use bluebird instead of Q? You can access it through Sequelize.Promise, it has all the methods like `Promise.all()` and many more aswell.

**Comment by jdnichollsc** [comment]
Ohh excellent with **bluebird**, but I have the same behavior using the following code

```javascript
models.sequelize.sync().then(function () {

    models.sequelize.Promise.mapSeries(forms, function (form) {
        return createForm(form);
    }).then(function(){
        console.log("Success!");
    }).catch(function(err){ //************ Doesn't work with one transaction **************
        console.log("Error!"); //Currently only works with more transactions
    });
});
```

Regards, Nicholls

**Comment by sushantdhiman** [comment]
https://github.com/sequelize/sequelize/pull/6319

**Comment by sushantdhiman** [comment]
@jdnichollsc We need to shim Promise methods because we use cls to maintain context in transactions. Here is the methods list https://github.com/sequelize/sequelize/blob/master/lib/promise.js#L27 which can be used safely with transactions

**Comment by jdnichollsc** [comment]
Hi @sushantdhiman 

Is not possible without indicate different transactions? so, it's necessary to use transactions with Promises?

_The error is a **ghost** without use different transactions_ :o

Thanks for your help, Nicholls

**Comment by sushantdhiman** [comment]
We need to pass certain data when running transactions. This is achieved by shimming bluebird promises with cls support.

So if you are using Transactions it is better use `Sequelize.Promise` (bluebird) instance @jdnichollsc

**Comment by felixfbecker** [comment]
I wish we would not use cls (at least internally). It has so much magic and monkey-patching involved. I don't get what's wrong with passing context explicitely as arguments/options.

**Comment by mickhansen** [comment]
@felixfbecker We don't use it internally unless the user "asks" us to.

**Comment by sushantdhiman** [comment]
@felixfbecker Exactly, but I am not ready to touch that code quite yet :wink: 

I think @overlookmotel introduced it, there must be a good reason to use them

**Comment by jdnichollsc** [comment]
Guys, check the error in my tests:

- Without indicate transactions: http://screencast.com/t/fFEqwaSZl5jj
- Using transactions: http://screencast.com/t/vIuPM7fFS3G

Thanks!

**Comment by sushantdhiman** [comment]
@jdnichollsc Thats what https://github.com/sequelize/sequelize/pull/6319 fixed (to support mapSeries in transactions), but its not released to `v3` yet. Its going to be released with `v4` :)

**Comment by jdnichollsc** [comment]
Ohh thanks for your help and sorry for my bad english hahaha

**Comment by overlookmotel** [comment]
It wasn't actually me that introduced CLS for transactions to Sequelize. I think credit goes to @mickhansen. I just thought it was fantastic and jumped right on using it.

@felixfbecker There's nothing wrong with passing context explicitly. It's probably a more robust way of doing things. But I found my code was getting so full of `{transaction: req.transaction, logging: req.log}` that it became hard to read. And then occasionally I'd miss one out by accident and get weird bugs. So for me, CLS is a godsend.

I'm still dragging on with my new `cls-bluebird` shim. It's almost there - written tests for every method except `coroutine` now. And along the way, I've uncovered a few bugs in `bluebird`, including [one pretty nasty one](https://github.com/petkaantonov/bluebird/issues/1148). Only a few hours work left, I'd say, and it'll be complete. But problem is I'm mad busy until end of month, so I doubt I'll get those hours until Sept.

I've also realised that there's a way to avoid using a patched `bluebird` internally within sequelize and still maintaining the ability to get transaction from CLS. But it'd require quite a lot of changes. If every sequelize public method at the start of the call (before any async action) checked the CLS namespace for `transaction`, and added it as `options.transaction`, then the transaction would be passed explicitly from that point on, and there'd be no need to ensure CLS context is maintained.

This would probably be a more robust solution than what we have now. But it'd take a bit of time and care to implement. But anyway, when my `cls-bluebird` patch is finally released, I'm pretty confident it'll be bulletproof!

**Comment by overlookmotel** [comment]
Incidentally, I haven't read the OP in great details, but I'm not 100% sure that #6319 would fix it - might be something else weird going on.

@jdnichollsc Could you apply the patch from #6319 (it's only two lines!) and check if that solves your problem?

**Comment by jdnichollsc** [comment]
The #6319 doesn't fix my issue :( 

Check the video: http://screencast.com/t/PkxFdU9U76

Regards, Nicholls

**Comment by sushantdhiman** [comment]
@jdnichollsc Can you try it with `Sequelize.Promise.all`, you are still using `Q.all`. Use only `Sequelize.Promise` in your code

**Comment by overlookmotel** [comment]
@jdnichollsc Going back to your very first example, does it output `Success!`?

I suspect this isn't actually a matter of the error being "swallowed" in the early example, but is due to the code behaving differently depending on whether you run it in transactions or not.

There are many uses of `findOrCreate()` in your examples, and also you have some parallel execution going on (`Q.all(form.fields.map(function(field){ ...`). Within a transaction, the `find()` part of `findOrCreate()` can only see database changes *made within its own transaction*, so it may `create()` the same row again. But then why the transactions try to commit, you end up with two rows which both can't be committed, creating the `UniqueConstraint` error.

Your are passing the transaction explicitly between calls, so CLS is not being used. So it should not make any difference whether you use `Q` or `Sequelize.Promise`. Feel free to use whichever you prefer.

What I'm saying is that at this point, I suspect there's no bug in Sequelize, and it's a problem with your code.

If you want further help here, I think you first need to try to narrow down this problem yourself. In particular, you haven't posted the data in `forms` so there's no way for anyone else reading to replicate the problem.

If I were you, I would try to reduce that data down piece by piece and find out at what point the error doesn't not occur any more. That I think will give you the clue what's going on.

If I'm wrong, and you do still think it's a bug in Sequelize, please come back with a single solid example of where an error is being thrown and you think it shouldn't be. That example should include: model definition, data being used (i.e. what is in the `fields.json` file), and the code - everything someone else needs to get the example running on their own computer and replicate the error. Personally, I find the screencasts hard to follow.

I hope you understand that I'm trying to help!

**Comment by jdnichollsc** [comment]
Hi @sushantdhiman 

Doesn't work using `Sequelize.Promise.all`, please reopen this issue!

Regards, Nicholls

**Comment by overlookmotel** [comment]
@jdnichollsc Please see my comment above. If you want help, you really need to give us more solid information than "it doesn't work".

> please come back with a single solid example of where an error is being thrown and you think it shouldn't be. That example should include: model definition, data being used (i.e. what is in the fields.json file), and the code - everything someone else needs to get the example running on their own computer and replicate the error. Personally, I find the screencasts hard to follow.

Please bear in mind that the maintainers work in their spare time. Help them to help you!

**Comment by jdnichollsc** [comment]
Hi guys,

Sorry for the delay, check my example to reproduce the issue

[issue.zip](https://github.com/sequelize/sequelize/files/556675/issue.zip)

Thanks for all and sorry again, Nicholls

**Comment by janmeier** [comment]
@jdnichollsc Please try to cut it down to the least amount of code possible, instead of posting your full project ....

[SSCCE](HTTP://SSCCE.ORG/)

**Comment by overlookmotel** [comment]
@jdnichollsc I'm going to put this less politely than @janmeier...

Are you bloody mad? I see you are very active on Github so I don't know how you haven't managed to understand this already.

Sending a massive load of stuff for people to wade through is rude. What it says is "My time is so important, I'm not going to bother putting any effort into this. Your time is worthless so you can spend an hour of it trying to figure out what's going on in my huge example".

I already gave you a long considered answer above working through your previous example as best I could with information you'd given. But you haven't bothered to engage or reply to that either.

In conclusion: you need to put in some work yourself to help people help you.

Or put another way: wake up or fuck off!

**Comment by overlookmotel** [comment]
Sorry I am in a really bad mood today and what I wrote above is probably inappropriate. But I see this happening all the time. How do people not understand to think of other people as well as themselves?

**Comment by jdnichollsc** [comment]
Incredible... this ticket only is a issue reported to help you guys to identify an error not caught by Sequelize about concurrency... and it's just an example, it isn't my full project... @janmeier

**Comment by janmeier** [comment]
Your zip files contain 1000 lines of .js code, I wouldn't exactly call that a short example. I'm pretty sure a lot can be trimmed down, while still showing the issue

**Comment by jdnichollsc** [comment]
Because the structure is created by Express Generator. The "./migrations/seed.js" file only have 84 lines of code, so I don't understand... what is the problem with that?

**Comment by jdnichollsc** [comment]
@janmeier The zip file was modified.

**Comment by felixfbecker** [comment]
@jdnichollsc Can't you just give us a single file? Use this template [here](https://github.com/sequelize/sequelize/blob/master/sscce_template.js)

**Comment by overlookmotel** [comment]
@jdnichollsc I'm sorry that I was rude before. But I do think the point I was making is valid.

Here's a quote from my previous comment above (which I put a fair amount of time and thought into):

> What I'm saying is that at this point, I suspect there's no bug in Sequelize, and it's a problem with your code.
> 
> If you want further help here, I think you first need to try to narrow down this problem yourself. In particular, you haven't posted the data in forms so there's no way for anyone else reading to replicate the problem.
> 
> If I were you, I would try to reduce that data down piece by piece and find out at what point the error doesn't not occur any more. That I think will give you the clue what's going on.
> 
> If I'm wrong, and you do still think it's a bug in Sequelize, please come back with a single solid example of where an error is being thrown and you think it shouldn't be. That example should include: model definition, data being used (i.e. what is in the fields.json file), and the code - everything someone else needs to get the example running on their own computer and replicate the error. Personally, I find the screencasts hard to follow.

Basically, I was saying that I don't think the problem is a bug in Sequelize. I think it's more likely a fault in the logic/code of your app. So when you said "this ticket only is an issue reported to help you guys to identify an error not caught by Sequelize about concurrency", I don't think that's true. And if you spent some time trying to solve this yourself, I think you'd discover that.

But I could be wrong! Hence why I said, if you can narrow it down and show that it *is* a bug in Sequelize then please come back to us and someone will try to fix asap.

The reason I was annoyed was because I spent time and thought trying to help you and you seemed to have totally ignore everything I'd said.
